File: src/Array.java

import java.util.Arrays;

@SuppressWarnings("unchecked")
public class Array<T> {
    private T[] items;
    private int count;

    public Array(int length) {
        items = (T[]) new Object[length];
    }

    public T[] getItems() {
        T[] newItems = (T[]) new Object[count];
        for (int i = 0; i < count; i++) {
            newItems[i] = items[i];
        }
        return newItems;
    }

    private void resizeIfRequired() {
        if (items.length == count) {
            T[] newItems = (T[]) new Object[count * 2];

            for (int i = 0; i < count; i++)
                newItems[i] = items[i];

            items = newItems;
        }
    }

    public void insert(T item) {
        resizeIfRequired();

        items[count++] = item;
    }

    public void insertAt(int index, T item) {
        if (index < 0 || index > count)
            throw new IllegalArgumentException();

        resizeIfRequired();

        for (int i = count - 1; i >= index; i--)
            items[i + 1] = items[i];

        items[index] = item;
        count++;
    }

    public void removeAt(int index) {
        if (index < 0 || index > count)
            throw new IllegalArgumentException();

        for (int i = index; i < count; i++) {
            var nextIndex = i + 1;
            if (items.length != nextIndex)
                items[i] = items[nextIndex];
        }

        count--;
    }

    public int indexOf(T item) {
        for (int i = 0; i < count; i++)
            if (item == items[i])
                return i;

        return -1;
    }

    public T max() {
        if (count == 0)
            return items[0];

        T max = items[0];

        for (T t : items) {
            if ((int) t > (int) max)
                max = t;
        }

        return max;
    }

    public T[] reverse() {
        T[] newItems = (T[]) new Object[count];

        for (int i = 0; i < count; i++)
            newItems[i] = items[count - i - 1];

        return newItems;
    }

    public Array<T> intersect(Array<T> other) {
        var intersection = new Array<T>(count);

        for (int i = 0; i < count; i++) {
            if (other.indexOf(items[i]) >= 0) {
                intersection.insert(items[i]);
            }
        }

        return intersection;
    }
}

class TestArray {
    public static void main(String[] args) {
        Array<Integer> numbers = new Array<>(1);

        numbers.insert(99);
        numbers.insert(2);
        numbers.insert(4);

        numbers.insertAt(2, 3);
        numbers.removeAt(0);

        System.out.println(Arrays.toString(numbers.getItems()));
        System.out.printf("reverse %s\n", Arrays.toString(numbers.reverse()));
        System.out.printf("index of `1` is %s.\n", Integer.toString(numbers.indexOf(1)));
        System.out.printf("index of `2` is %s.\n", Integer.toString(numbers.indexOf(2)));
        System.out.printf("index of `4` is %s.\n", Integer.toString(numbers.indexOf(4)));
        System.out.printf("max %s.\n", Integer.toString(numbers.max()));

        Array<Integer> moreNumbers = new Array<>(4);

        moreNumbers.insert(4);
        moreNumbers.insert(5);

        System.out.println(Arrays.toString(numbers.intersect(moreNumbers).getItems()));
    }
}

File: cpp/src/array.cpp
#include "array.hpp"

Array::Array(int length) : length(length), items(new int[length]), count(0) {}

Array::~Array() { delete[] items; }

void Array::insert(int item) {
  if (this->count >= length) {
    length *= 2;

    auto newItems = new int[length];

    for (int i = 0; i < this->count; i++) {
      newItems[i] = this->items[i];
    }

    delete[] this->items;
    this->items = nullptr;

    this->items = newItems;
  }

  this->items[this->count] = item;

  this->count++;
}

void Array::insert_at(int item, int index) {
  if (this->count == 0 && index == 0) {
    this->insert(item);
    return;
  }

  if (index < 0 || index >= this->count) {
    return;
  }

  auto item_to_insert = item;

  for (int i = index; i <= this->count; i++) {
    if (i == this->count) {
      this->insert(item_to_insert);
      break;
    } else {
      auto temp = this->items[i];
      this->items[i] = item_to_insert;
      item_to_insert = temp;
    }
  }
}

void Array::remove_at(int index) {
  if (index < 0 || index >= this->count) {
    throw std::out_of_range("Index out of bounds");
  }

  for (int i = index; i < this->count; i++) {
    this->items[i] = this->items[i + 1];
  }

  this->count--;
}

void Array::reverse() {
  auto newItems = new int[this->count];
  for (int i = 0; i < this->count; i++) {
    newItems[i] = this->items[this->count - i - 1];
  }

  delete[] this->items;
  this->items = newItems;
}

int Array::indexof(int item) const {
  for (int i = 0; i < this->count; i++) {
    if (this->items[i] == item) {
      return i;
    }
  }

  return -1;
}

int Array::max_index() const {
  if (this->count == 0) {
    return -1;
  }

  auto idx = 0;
  auto maxElement = this->items[idx];
  for (int i = 0; i < this->count; i++) {
    auto element = this->items[i];
    if (element > maxElement) {
      maxElement = element;
      idx = i;
    }
  }

  return idx;
}

Array* Array::intersect(const Array& other) const {
  auto array = new Array(this->count);

  for (int i = 0; i < this->count; i++) {
    if (other.indexof(this->items[i]) == -1) {
      continue;
    }

    array->insert(this->items[i]);
  }

  return array;
}

std::ostream& operator<<(std::ostream& stream, const Array& array) {
  for (int i = 0; i < array.count; i++) {
    stream << array.items[i];
  }

  return stream;
}
